src/goto-symex/symex_goto.cpp

```diff
@@ -469,17 +469,23 @@ static void merge_names(
   //  1. Either guard is false, so we can't follow that branch.
   //  2. Either identifier is of generation zero, and so hasn't been
   //     initialized and therefore an invalid target.
-  if(dest_state.guard.is_false())
-    rhs = goto_state_rhs;
-  else if(goto_state.guard.is_false())
-    rhs = dest_state_rhs;
-  else if(goto_count == 0)
+  if(
+    dest_state.guard.is_false() ||
+    (dest_count == 0 && !symbol.value.is_not_nil()))
   {
-    rhs = dest_state_rhs;
+    if(goto_count == 0)
```

Comment:
Note the `dest_count == 0 && goto_count == 0` case is forbidden above, so this implies `dest_state.guard.is_false() && dest_count != 0) -- not sure if this can be distangled to make that clearer, or perhaps clarify with a comment

Reply:
If `dest_state.guard.is_false()` then really we don't care about `dest_count`, because it sits on a dead branch. Any idea what you'd like the comment to say?

Reply:
How about just pulling out

```
if(
  (dest_state.guard.is_false() && goto_count == 0) ||
  (goto_state.guard.is_false() && dest_count == 0))
{
  return;
}
```

It does mean testing `is_false()` multiple times, but it makes it clearer which cases are possible. Comment-wise, I guess we should add something like `// If the symbol is not initialised on one path and the other path is unreachable we don't need to generate anything`